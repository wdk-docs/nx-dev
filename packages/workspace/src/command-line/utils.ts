import * as yargsParser from 'yargs-parser';
import * as yargs from 'yargs';
import { readNxJson } from '../core/file-utils';
import { output } from '../utilities/output';
import { names } from '@nrwl/devkit';
import type { NxAffectedConfig } from '@nrwl/devkit';

const runOne: string[] = [
  'target',
  'configuration',
  'prod',
  'runner',
  'parallel',
  'max-parallel',
  'exclude',
  'only-failed',
  'help',
  'with-deps',
  'skip-nx-cache',
  'scan',
  'hide-cached-output',
];

const runMany: string[] = [...runOne, 'projects', 'all'];

const runAffected: string[] = [
  ...runOne,
  'untracked',
  'uncommitted',
  'all',
  'base',
  'head',
  'files',
  'plain',
  'select',
];

export interface RawNxArgs extends NxArgs {
  prod?: boolean;
}

export interface NxArgs {
  target?: string;
  configuration?: string;
  runner?: string;
  parallel?: number;
  untracked?: boolean;
  uncommitted?: boolean;
  all?: boolean;
  base?: string;
  head?: string;
  exclude?: string[];
  files?: string[];
  onlyFailed?: boolean;
  'only-failed'?: boolean;
  verbose?: boolean;
  help?: boolean;
  version?: boolean;
  plain?: boolean;
  withDeps?: boolean;
  'with-deps'?: boolean;
  projects?: string[];
  select?: string;
  skipNxCache?: boolean;
  'skip-nx-cache'?: boolean;
  'hide-cached-output'?: boolean;
  hideCachedOutput?: boolean;
  scan?: boolean;
}

const ignoreArgs = ['$0', '_'];

export function splitArgsIntoNxArgsAndOverrides(
  args: yargs.Arguments,
  mode: 'run-one' | 'run-many' | 'affected' | 'print-affected',
  options = { printWarnings: true }
): { nxArgs: NxArgs; overrides: yargs.Arguments } {
  const nxSpecific =
    mode === 'run-one' ? runOne : mode === 'run-many' ? runMany : runAffected;

  const nxArgs: RawNxArgs = {};
  const overrides = yargsParser(args._ as string[], {
    configuration: {
      'strip-dashed': true,
      'dot-notation': false,
    },
  });
  // This removes the overrides from the nxArgs._
  args._ = overrides._;

  delete overrides._;

  Object.entries(args).forEach(([key, value]) => {
    const dasherized = names(key).fileName;
    if (nxSpecific.includes(dasherized) || dasherized.startsWith('nx-')) {
      if (value !== undefined) nxArgs[key] = value;
    } else if (!ignoreArgs.includes(key)) {
      overrides[key] = value;
    }
  });

  if (mode === 'run-many') {
    if (!nxArgs.projects) {
      nxArgs.projects = [];
    } else {
      nxArgs.projects = (args.projects as string)
        .split(',')
        .map((p: string) => p.trim());
    }
  }

  if (nxArgs.prod) {
    delete nxArgs.prod;
    nxArgs.configuration = 'production';
  }

  if (mode === 'affected') {
    if (options.printWarnings && nxArgs.all) {
      output.warn({
        title: `Running affected:* commands with --all can result in very slow builds.`,
        bodyLines: [
          `${output.bold(
            '--all'
          )} is not meant to be used for any sizable project or to be used in CI.`,
          '',
          `${output.colors.gray(
            'Learn more about checking only what is affected: '
          )}https://nx.dev/latest/angular/cli/affected#affected.`,
        ],
      });
    }

    if (
      !nxArgs.files &&
      !nxArgs.uncommitted &&
      !nxArgs.untracked &&
      !nxArgs.base &&
      !nxArgs.head &&
      !nxArgs.all &&
      args._.length >= 3
    ) {
      nxArgs.base = args._[1] as string;
      nxArgs.head = args._[2] as string;
    }

    // Allow setting base and head via environment variables (lower priority then direct command arguments)
    if (!nxArgs.base && process.env.NX_BASE) {
      nxArgs.base = process.env.NX_BASE;
      if (options.printWarnings) {
        output.note({
          title: `No explicit --base argument provided, but found environment variable NX_BASE so using its value as the affected base: ${output.bold(
            `${nxArgs.base}`
          )}`,
        });
      }
    }
    if (!nxArgs.head && process.env.NX_HEAD) {
      nxArgs.head = process.env.NX_HEAD;
      if (options.printWarnings) {
        output.note({
          title: `No explicit --head argument provided, but found environment variable NX_HEAD so using its value as the affected head: ${output.bold(
            `${nxArgs.head}`
          )}`,
        });
      }
    }

    if (!nxArgs.base) {
      const affectedConfig = getAffectedConfig();
      nxArgs.base = affectedConfig.defaultBase;

      // No user-provided arguments to set the affected criteria, so inform the user of the defaults being used
      if (
        options.printWarnings &&
        !nxArgs.head &&
        !nxArgs.files &&
        !nxArgs.uncommitted &&
        !nxArgs.untracked &&
        !nxArgs.all
      ) {
        output.note({
          title: `Affected criteria defaulted to --base=${output.bold(
            `${nxArgs.base}`
          )} --head=${output.bold('HEAD')}`,
        });
      }
    }
  }

  if (!nxArgs.skipNxCache) {
    nxArgs.skipNxCache = process.env.NX_SKIP_NX_CACHE === 'true';
  }

  if (args['parallel'] === 'false' || args['parallel'] === false) {
    nxArgs['parallel'] = 1;
  } else if (
    args['parallel'] === 'true' ||
    args['parallel'] === true ||
    args['parallel'] === ''
  ) {
    nxArgs['parallel'] = Number(
      nxArgs['maxParallel'] || nxArgs['max-parallel'] || 3
    );
  } else if (args['parallel'] !== undefined) {
    nxArgs['parallel'] = Number(args['parallel']);
  }

  return { nxArgs, overrides };
}

export function getAffectedConfig(): NxAffectedConfig {
  const config = readNxJson();

  return {
    defaultBase: config.affected?.defaultBase || 'main',
  };
}
