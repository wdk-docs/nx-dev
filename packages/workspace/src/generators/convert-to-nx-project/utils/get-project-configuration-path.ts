import { ProjectConfiguration } from '@nrwl/devkit';
import { join } from 'path';

export function getProjectConfigurationPath(
  configuration: ProjectConfiguration
) {
  return join(configuration.root, 'project.json');
}
