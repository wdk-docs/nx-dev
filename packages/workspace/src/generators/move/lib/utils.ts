import {
  getWorkspaceLayout,
  joinPathFragments,
  ProjectConfiguration,
  Tree,
} from '@nrwl/devkit';

import { Schema } from '../schema';

/**
 * This helper function ensures that we don't move libs or apps
 * outside of the folders they should be in.
 *
 * This will break if someone isn't using the default libs/apps
 * folders. In that case, they're on their own :/
 */
export function getDestination(
  host: Tree,
  schema: Schema,
  project: ProjectConfiguration
): string {
  const projectType = project.projectType;

  const workspaceLayout = getWorkspaceLayout(host);

  let rootFolder = workspaceLayout.libsDir;
  if (projectType === 'application') {
    rootFolder = workspaceLayout.appsDir;
  }
  return joinPathFragments(rootFolder, schema.destination);
}

/**
 * Replaces slashes with dashes
 *
 * @param path
 */
export function getNewProjectName(path: string): string {
  return path.replace(/\//g, '-');
}

/**
 * Normalizes slashes (removes duplicates)
 *
 * @param input
 */
export function normalizeSlashes(input: string): string {
  return input
    .split('/')
    .filter((x) => !!x)
    .join('/');
}
