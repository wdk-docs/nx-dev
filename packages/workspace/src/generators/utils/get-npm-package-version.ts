export function getNpmPackageVersion(packageName: string): string | null {
  try {
    const version = require('child_process').execSync(
      `npm view ${packageName} version`,
      { stdio: ['pipe', 'pipe', 'ignore'] }
    );

    if (version) {
      return version
        .toString()
        .trim()
        .replace(/^\n*|\n*$/g, '');
    }
  } catch (err) {}
  return null;
}
