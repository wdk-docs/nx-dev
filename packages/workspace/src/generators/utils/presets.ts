export enum Preset {
  Apps = 'apps',
  Empty = 'empty', // same as apps, deprecated
  Core = 'core',
  NPM = 'npm', // same as core, deprecated
  TS = 'ts',
  WebComponents = 'web-components',
  Angular = 'angular',
  AngularWithNest = 'angular-nest',
  React = 'react',
  ReactWithExpress = 'react-express',
  ReactNative = 'react-native',
  NextJs = 'next',
  Nest = 'nest',
  Express = 'express',
}
