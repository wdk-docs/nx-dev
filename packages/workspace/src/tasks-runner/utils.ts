import {
  getPackageManagerCommand,
  ProjectGraph,
  ProjectGraphNode,
  TargetDependencyConfig,
  Task,
  TaskGraph,
} from '@nrwl/devkit';
import { flatten } from 'flat';
import { output } from '../utilities/output';
import { Workspaces } from '@nrwl/tao/src/shared/workspace';
import { mergeNpmScriptsWithTargets } from '../utilities/project-graph-utils';
import { existsSync } from 'fs';
import { join } from 'path';
import {
  loadNxPlugins,
  mergePluginTargetsWithNxTargets,
} from '@nrwl/tao/src/shared/nx-plugin';

export function getCommandAsString(task: Task) {
  const execCommand = getPackageManagerCommand().exec;
  const args = getCommandArgsForTask(task);
  return [execCommand, 'nx', ...args].join(' ').trim();
}

export function getDependencyConfigs(
  { project, target }: { project: string; target: string },
  defaultDependencyConfigs: Record<string, TargetDependencyConfig[]>,
  projectGraph: ProjectGraph
): TargetDependencyConfig[] | undefined {
  // DependencyConfigs configured in workspace.json override configurations at the root.
  const dependencyConfigs =
    projectGraph.nodes[project].data?.targets[target]?.dependsOn ??
    defaultDependencyConfigs[target] ??
    [];

  for (const dependencyConfig of dependencyConfigs) {
    if (
      dependencyConfig.projects !== 'dependencies' &&
      dependencyConfig.projects !== 'self'
    ) {
      output.error({
        title: `dependsOn is improperly configured for ${project}:${target}`,
        bodyLines: [
          `dependsOn.projects is ${dependencyConfig.projects} but should be "self" or "dependencies"`,
        ],
      });
      process.exit(1);
    }
  }
  return dependencyConfigs;
}

export function getOutputs(p: Record<string, ProjectGraphNode>, task: Task) {
  return getOutputsForTargetAndConfiguration(task, p[task.target.project]);
}

export function getOutputsForTargetAndConfiguration(
  task: Pick<Task, 'target' | 'overrides'>,
  node: ProjectGraphNode
) {
  const { target, configuration } = task.target;

  const targets = node.data.targets[target];

  const options = {
    ...targets.options,
    ...targets?.configurations?.[configuration],
    ...task.overrides,
  };

  if (targets?.outputs) {
    return targets.outputs
      .map((output: string) => interpolateOutputs(output, options))
      .filter((output) => !!output);
  }

  // Keep backwards compatibility in case `outputs` doesn't exist
  if (options.outputPath) {
    return Array.isArray(options.outputPath)
      ? options.outputPath
      : [options.outputPath];
  } else if (target === 'build' || target === 'prepare') {
    return [
      `dist/${node.data.root}`,
      `${node.data.root}/dist`,
      `${node.data.root}/build`,
      `${node.data.root}/public`,
    ];
  } else {
    return [];
  }
}

export function unparse(options: Object): string[] {
  const unparsed = [];
  for (const key of Object.keys(options)) {
    const value = options[key];
    unparseOption(key, value, unparsed);
  }

  return unparsed;
}

function unparseOption(key: string, value: any, unparsed: string[]) {
  if (value === true) {
    unparsed.push(`--${key}`);
  } else if (value === false) {
    unparsed.push(`--no-${key}`);
  } else if (Array.isArray(value)) {
    value.forEach((item) => unparseOption(key, item, unparsed));
  } else if (Object.prototype.toString.call(value) === '[object Object]') {
    const flattened = flatten<any, any>(value, { safe: true });
    for (const flattenedKey in flattened) {
      unparseOption(
        `${key}.${flattenedKey}`,
        flattened[flattenedKey],
        unparsed
      );
    }
  } else if (
    typeof value === 'string' &&
    stringShouldBeWrappedIntoQuotes(value)
  ) {
    const sanitized = value.replace(/"/g, String.raw`\"`);
    unparsed.push(`--${key}="${sanitized}"`);
  } else if (value != null) {
    unparsed.push(`--${key}=${value}`);
  }
}

function stringShouldBeWrappedIntoQuotes(str: string) {
  return str.includes(' ') || str.includes('{') || str.includes('"');
}

function interpolateOutputs(template: string, data: any): string {
  return template.replace(/{([\s\S]+?)}/g, (match: string) => {
    let value = data;
    let path = match.slice(1, -1).trim().split('.').slice(1);
    for (let idx = 0; idx < path.length; idx++) {
      if (!value[path[idx]]) {
        return;
      }
      value = value[path[idx]];
    }

    return value;
  });
}

export function getExecutorNameForTask(task: Task, workspace: Workspaces) {
  const workspaceConfiguration = workspace.readWorkspaceConfiguration();
  const project = workspaceConfiguration.projects[task.target.project];

  if (existsSync(join(project.root, 'package.json'))) {
    project.targets = mergeNpmScriptsWithTargets(project.root, project.targets);
  }
  project.targets = mergePluginTargetsWithNxTargets(
    project.root,
    project.targets,
    loadNxPlugins(workspaceConfiguration.plugins)
  );

  if (!project.targets[task.target.target]) {
    throw new Error(
      `Cannot find configuration for task ${task.target.project}:${task.target.target}`
    );
  }

  return project.targets[task.target.target].executor;
}

export function getExecutorForTask(task: Task, workspace: Workspaces) {
  const executor = getExecutorNameForTask(task, workspace);
  const [nodeModule, executorName] = executor.split(':');

  return workspace.readExecutor(nodeModule, executorName);
}

export function getCustomHasher(task: Task, workspace: Workspaces) {
  try {
    const factory = getExecutorForTask(task, workspace).hasherFactory;
    return factory ? factory() : null;
  } catch (e) {
    console.error(e);
    throw new Error(`Unable to load hasher for task "${task.id}"`);
  }
}

export function removeTasksFromTaskGraph(
  graph: TaskGraph,
  ids: string[]
): TaskGraph {
  const tasks = {};
  const dependencies = {};
  const removedSet = new Set(ids);
  for (let taskId of Object.keys(graph.tasks)) {
    if (!removedSet.has(taskId)) {
      tasks[taskId] = graph.tasks[taskId];
      dependencies[taskId] = graph.dependencies[taskId].filter(
        (depTaskId) => !removedSet.has(depTaskId)
      );
    }
  }
  return {
    tasks,
    dependencies: dependencies,
    roots: Object.keys(dependencies).filter(
      (k) => dependencies[k].length === 0
    ),
  };
}

export function calculateReverseDeps(
  taskGraph: TaskGraph
): Record<string, string[]> {
  const reverseTaskDeps: Record<string, string[]> = {};
  Object.keys(taskGraph.tasks).forEach((t) => {
    reverseTaskDeps[t] = [];
  });

  Object.keys(taskGraph.dependencies).forEach((taskId) => {
    taskGraph.dependencies[taskId].forEach((d) => {
      reverseTaskDeps[d].push(taskId);
    });
  });

  return reverseTaskDeps;
}

export function getCliPath(workspaceRoot: string) {
  const cli = require.resolve(`@nrwl/cli/lib/run-cli.js`, {
    paths: [workspaceRoot],
  });
  return `${cli}`;
}

export function getCommandArgsForTask(task: Task) {
  const args: string[] = unparse(task.overrides || {});

  const target = task.target.target.includes(':')
    ? `"${task.target.target}"`
    : task.target.target;

  const config = task.target.configuration
    ? `:${task.target.configuration}`
    : '';

  return ['run', `${task.target.project}:${target}${config}`, ...args];
}

export function shouldForwardOutput(
  task: Task,
  initiatingProject: string | null,
  options: {
    cacheableOperations?: string[] | null;
    cacheableTargets?: string[] | null;
  }
): boolean {
  if (process.env.NX_FORWARD_OUTPUT === 'true') return true;
  if (!isCacheableTask(task, options)) return true;
  if (task.target.project === initiatingProject) return true;
  return false;
}

export function isCacheableTask(
  task: Task,
  options: {
    cacheableOperations?: string[] | null;
    cacheableTargets?: string[] | null;
  }
): boolean {
  const cacheable = options.cacheableOperations || options.cacheableTargets;
  return (
    cacheable &&
    cacheable.indexOf(task.target.target) > -1 &&
    !longRunningTask(task)
  );
}

function longRunningTask(task: Task) {
  return !!task.overrides['watch'];
}
