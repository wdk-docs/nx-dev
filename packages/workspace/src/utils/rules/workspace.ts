import { JsonObject } from '@angular-devkit/core';
import { Rule } from '@angular-devkit/schematics';
import { updateWorkspace } from '../workspace';

/**
 * Sets the default collection to the provided collection name
 * The collection name is only set in case the current defaultCollection is undefined or set to '@nrwl/workspace'
 * @param collectionName Name of the collection to be set as defaultCollection
 */
export function setDefaultCollection(collectionName: string): Rule {
  return updateWorkspace((workspace) => {
    workspace.extensions.cli = workspace.extensions.cli || {};

    const defaultCollection: string =
      workspace.extensions.cli &&
      ((workspace.extensions.cli as JsonObject).defaultCollection as string);

    if (!defaultCollection || defaultCollection === '@nrwl/workspace') {
      (workspace.extensions.cli as JsonObject).defaultCollection =
        collectionName;
    }
  });
}
