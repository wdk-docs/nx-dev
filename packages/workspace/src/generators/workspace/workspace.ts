import {
  generateFiles,
  Tree,
  updateJson,
  convertNxGenerator,
  names,
  writeJson,
  formatFiles,
} from '@nrwl/devkit';
import { Schema } from './schema';
import {
  angularCliVersion,
  nxVersion,
  prettierVersion,
  typescriptVersion,
} from '../../utils/versions';
import { readFileSync } from 'fs';
import { join, join as pathJoin } from 'path';
import { reformattedWorkspaceJsonOrNull } from '@nrwl/tao/src/shared/workspace';
import { Preset } from '../utils/presets';
import { deduceDefaultBase } from '../../utilities/default-base';

export const DEFAULT_NRWL_PRETTIER_CONFIG = {
  singleQuote: true,
};

function decorateAngularClI(host: Tree, options: Schema) {
  const decorateCli = readFileSync(
    pathJoin(__dirname as any, '..', 'utils', 'decorate-angular-cli.js__tmpl__')
  ).toString();
  host.write(join(options.directory, 'decorate-angular-cli.js'), decorateCli);
}

function setPresetProperty(tree: Tree, options: Schema) {
  updateJson(tree, join(options.directory, 'nx.json'), (json) => {
    if (
      options.preset === Preset.Core ||
      options.preset === Preset.TS ||
      options.preset === Preset.NPM
    ) {
      addPropertyWithStableKeys(
        json,
        'extends',
        '@nrwl/workspace/presets/core.json'
      );
      delete json.implicitDependencies;
      delete json.targetDependencies;
      delete json.workspaceLayout;
    }
    return json;
  });
}

function createAppsAndLibsFolders(host: Tree, options: Schema) {
  if (
    options.preset === Preset.Core ||
    options.preset === Preset.TS ||
    options.preset === Preset.NPM
  ) {
    host.write(join(options.directory, 'packages/.gitkeep'), '');
  } else {
    host.write(join(options.directory, 'apps/.gitkeep'), '');
    host.write(join(options.directory, 'libs/.gitkeep'), '');
  }
}

function createFiles(host: Tree, options: Schema) {
  const npmScope = options.npmScope ?? options.name;
  const formattedNames = names(options.name);
  generateFiles(host, pathJoin(__dirname, './files'), options.directory, {
    formattedNames,
    dot: '.',
    tmpl: '',
    workspaceFile: options.cli === 'angular' ? 'angular' : 'workspace',
    cliCommand: options.cli === 'angular' ? 'ng' : 'nx',
    nxCli: false,
    typescriptVersion,
    prettierVersion,
    // angular cli is used only when workspace schematics is added to angular cli
    angularCliVersion,
    ...(options as object),
    nxVersion,
    npmScope,
    packageManager: options.packageManager,
  });
}

function createPrettierrc(host: Tree, options: Schema) {
  writeJson(
    host,
    join(options.directory, '.prettierrc'),
    DEFAULT_NRWL_PRETTIER_CONFIG
  );
}

function formatWorkspaceJson(host: Tree, options: Schema) {
  const path = join(
    options.directory,
    options.cli === 'angular' ? 'angular.json' : 'workspace.json'
  );

  try {
    updateJson(host, path, (workspaceJson) => {
      const reformatted = reformattedWorkspaceJsonOrNull(workspaceJson);
      if (reformatted) {
        return reformatted;
      }
      return workspaceJson;
    });
  } catch (e) {
    console.error(`Failed to format: ${path}`);
    console.error(e);
  }
}

function addNpmScripts(host: Tree, options: Schema) {
  if (options.cli === 'angular') {
    updateJson(host, join(options.directory, 'package.json'), (json) => {
      Object.assign(json.scripts, {
        ng: 'nx',
        postinstall: 'node ./decorate-angular-cli.js',
      });
      return json;
    });
  }

  if (
    options.preset !== Preset.TS &&
    options.preset !== Preset.Core &&
    options.preset !== Preset.NPM
  ) {
    updateJson(host, join(options.directory, 'package.json'), (json) => {
      Object.assign(json.scripts, {
        start: 'nx serve',
        build: 'nx build',
        test: 'nx test',
      });
      return json;
    });
  }
}

export async function workspaceGenerator(host: Tree, options: Schema) {
  if (!options.name) {
    throw new Error(`Invalid options, "name" is required.`);
  }
  options = normalizeOptions(options);
  createFiles(host, options);
  createPrettierrc(host, options);
  if (options.cli === 'angular') {
    decorateAngularClI(host, options);
  }
  setPresetProperty(host, options);
  addNpmScripts(host, options);
  createAppsAndLibsFolders(host, options);

  await formatFiles(host);
  formatWorkspaceJson(host, options);
}

export const workspaceSchematic = convertNxGenerator(workspaceGenerator);

function addPropertyWithStableKeys(obj: any, key: string, value: string) {
  const copy = { ...obj };
  Object.keys(obj).forEach((k) => {
    delete obj[k];
  });
  obj[key] = value;
  Object.keys(copy).forEach((k) => {
    obj[k] = copy[k];
  });
}

function normalizeOptions(options: Schema) {
  let defaultBase = options.defaultBase || deduceDefaultBase();
  return {
    ...options,
    defaultBase,
  };
}
