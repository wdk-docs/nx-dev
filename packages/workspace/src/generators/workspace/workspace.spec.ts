import { readJson } from '@nrwl/devkit';
import type { Tree, NxJsonConfiguration } from '@nrwl/devkit';
import { workspaceGenerator } from './workspace';
import { createTree } from '@nrwl/devkit/testing';
import { Preset } from '../utils/presets';

describe('@nrwl/workspace:workspace', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTree();
  });

  it('should create files', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'nx',
      preset: Preset.Empty,
      defaultBase: 'main',
    });
    expect(tree.exists('/proj/nx.json')).toBe(true);
    expect(tree.exists('/proj/workspace.json')).toBe(true);
    expect(tree.exists('/proj/.prettierrc')).toBe(true);
    expect(tree.exists('/proj/.prettierignore')).toBe(true);
  });

  it('should create nx.json', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'nx',
      preset: Preset.Empty,
      defaultBase: 'main',
    });
    const nxJson = readJson<NxJsonConfiguration>(tree, '/proj/nx.json');
    expect(nxJson).toEqual({
      npmScope: 'proj',
      affected: {
        defaultBase: 'main',
      },
      cli: {
        defaultCollection: '@nrwl/workspace',
      },
      implicitDependencies: {
        'package.json': {
          dependencies: '*',
          devDependencies: '*',
        },
        '.eslintrc.json': '*',
      },
      tasksRunnerOptions: {
        default: {
          runner: '@nrwl/workspace/tasks-runners/default',
          options: {
            cacheableOperations: ['build', 'lint', 'test', 'e2e'],
          },
        },
      },
      targetDependencies: {
        build: [
          {
            projects: 'dependencies',
            target: 'build',
          },
        ],
      },
    });
  });

  it('should create a prettierrc file', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'nx',
      preset: Preset.Empty,
      defaultBase: 'main',
    });
    expect(tree.read('proj/.prettierrc', 'utf-8')).toMatchSnapshot();
  });

  it('should recommend vscode extensions', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'nx',
      preset: Preset.Empty,
      defaultBase: 'main',
    });
    const recommendations = readJson<{ recommendations: string[] }>(
      tree,
      'proj/.vscode/extensions.json'
    ).recommendations;

    expect(recommendations).toMatchSnapshot();
  });

  it('should recommend vscode extensions (angular)', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'angular',
      preset: Preset.Empty,
      defaultBase: 'main',
    });
    const recommendations = readJson<{ recommendations: string[] }>(
      tree,
      'proj/.vscode/extensions.json'
    ).recommendations;

    expect(recommendations).toMatchSnapshot();
  });

  it('should add decorate-angular-cli when used with angular cli', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'angular',
      preset: Preset.Empty,
      defaultBase: 'main',
    });
    expect(tree.exists('/proj/decorate-angular-cli.js')).toBe(true);

    const { scripts } = readJson(tree, '/proj/package.json');
    expect(scripts).toMatchInlineSnapshot(`
Object {
  "build": "nx build",
  "ng": "nx",
  "postinstall": "node ./decorate-angular-cli.js",
  "start": "nx serve",
  "test": "nx test",
}
`);
  });

  it('should not add decorate-angular-cli when used with nx cli', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'nx',
      preset: Preset.Empty,
      defaultBase: 'main',
    });
    expect(tree.exists('/proj/decorate-angular-cli.js')).toBe(false);

    const { scripts } = readJson(tree, '/proj/package.json');
    expect(scripts).toMatchInlineSnapshot(`
Object {
  "build": "nx build",
  "start": "nx serve",
  "test": "nx test",
}
`);
  });

  it('should create a workspace using package layout', async () => {
    await workspaceGenerator(tree, {
      name: 'proj',
      directory: 'proj',
      cli: 'nx',
      preset: Preset.NPM,
      defaultBase: 'main',
    });
    expect(tree.exists('/proj/packages/.gitkeep')).toBe(true);
    expect(tree.exists('/proj/apps/.gitkeep')).toBe(false);
    expect(tree.exists('/proj/libs/.gitkeep')).toBe(false);
    const nx = readJson(tree, '/proj/nx.json');
    expect(nx.extends).toEqual('@nrwl/workspace/presets/core.json');

    const { scripts } = readJson(tree, '/proj/package.json');
    expect(scripts).toMatchInlineSnapshot(`Object {}`);
  });
});
