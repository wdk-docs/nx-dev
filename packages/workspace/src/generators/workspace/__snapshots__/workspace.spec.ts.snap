// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nrwl/workspace:workspace should create a prettierrc file 1`] = `
"{
  \\"singleQuote\\": true
}
"
`;

exports[`@nrwl/workspace:workspace should recommend vscode extensions (angular) 1`] = `
Array [
  "angular.ng-template",
  "nrwl.angular-console",
  "esbenp.prettier-vscode",
]
`;

exports[`@nrwl/workspace:workspace should recommend vscode extensions 1`] = `
Array [
  "nrwl.angular-console",
  "esbenp.prettier-vscode",
]
`;
